### ISSUE TITLE
missing CFG edges from return statement to enclosing finally blocks

### ISSUE THREAD

**Issue by msridhar** [typetools/checker-framework#1797]
I notice from [this comment](https://github.com/typetools/checker-framework/blob/4c7fa25c57f251d473d86ff74985d6d240e94e33/dataflow/src/org/checkerframework/dataflow/cfg/CFGBuilder.java#L4151) that there seem to be missing control-flow edges from a `return` statement to the beginning of enclosing finally blocks.  We noticed some unsoundness in NullAway that I think is due to this issue.  E.g., this example gets through NullAway with no warning:
```java
public void foo(@Nullable Object o) {
  try {
    return;
  } finally {
    o.toString();
  }
}
```
Looking under the hood, it seems the dataflow analysis does not even visit the `finally` block, I'm guessing because it starts from the entry and there's no path to the finally block from the entry block in the CFG?

I notice that the above example and other similar examples seem to be checked correctly by the Checker Framework's nullness checker.  Is there some special workaround in that checker that we can borrow?  Is the underlying issue hard to fix?

**Comment by msridhar** [comment]
@wmdietl I think this is actually a distinct issue from #293.  I read #293 as being about imprecision / over-approximation in the control-flow graph; both the "fall-through" and exceptional exits of a `try` block go to the same `finally` successor block (rather than distinct copies), so dataflow results from these cases are conflated.  For the case here, there is simply no control-flow edge from the `return` statement CFG node to the `finally` CFG node at all.  So when we compute a forward dataflow analysis, our analyzer never even analyzes the code in the `finally` block, as it looks unreachable.

/cc @lazaroclapp

**Comment by wmdietl** [comment]
Thanks for the update, I'll take a look.

**Comment by wmdietl** [comment]
@msridhar @lazaroclapp 
This turns out to be a bigger issue and I produced test cases that resulted in false positives, because dataflow refinement didn't happen for finally blocks at all.
I attempted a fix in #1814, see tests in https://github.com/typetools/checker-framework/blob/3efbd5d72f991b37cf771f5b2261a92f605d602c/checker/tests/nullness/Issue1797.java

I do notice that some CFG edges still look incorrect, but at least all finally blocks should now be reachable.
#1814 should be a step in the right direction.

Once #1814 is merged, I will revisit #293. I will finish the tree-cloning approach to see whether that can work nicely. If not, we can investigate the proposal from https://github.com/typetools/checker-framework/issues/293#issuecomment-362720562
Please let me know what you think of #1814 and this plan.

**Comment by msridhar** [comment]
The test cases look good!  And by good I mean nasty 😄 Not sure I'll have time to dig deep on the CFG construction changes; maybe @lazaroclapp can take a quick look.  My only thought is this change might affect other flow-sensitive type systems as well, so it might be good to have a couple test cases for those systems, as a sanity check.  Otherwise the plan sounds good.

**Comment by lazaroclapp** [comment]
For the record, #1814 makes sense to me as a fix. Haven't tested it yet, but both the test cases and the changed CFG code look good as far as I can tell. Thanks, @wmdietl !
